waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -246,4 +254,143 @@ proc decryptNoisePublicKey*(cs: ChaChaPolyCipherState, noisePublicKey: NoisePubl
+  return payload2
+
+
+# Serializes a PayloadV2 object to a byte sequences according to https://rfc.vac.dev/spec/35/.
+# The results can be passed to the payload field of a WakuMessage https://rfc.vac.dev/spec/14/
+proc serializePayloadV2*(self: PayloadV2): Result[seq[byte], cstring] 
+  {.raises: [Defect, NoiseMalformedHandshake, NoisePublicKeyError].} =
+
+  #We collect public keys contained in the handshake message
+  var
+    # According to https://rfc.vac.dev/spec/35/, the maximum size for the handshake message is 256 bytes, that is 
+    # the handshake message length can be represented with 1 byte only. (its length can be stored in 1 byte)
+    # However, to ease public keys length addition operation, we declare it as int and later cast to uit8
+    serializedHandshakeMessageLen: int = 0
+    # This variables will store the concatenation of the serializations of all public keys in the handshake message
+    serializedHandshakeMessage = newSeqOfCap[byte](256)
+    # A variable to store the currently processed public key serialization
+    serializedPk: seq[byte]
+  # For each public key in the handshake message
+  for pk in self.handshakeMessage:
+    # We serialize the public key
+    serializedPk = serializeNoisePublicKey(pk)
+    # We sum its serialized length to the total
+    serializedHandshakeMessageLen +=  serializedPk.len
+    # We add its serialization to the concatenation of all serialized public keys in the handshake message 
+    serializedHandshakeMessage.add serializedPk
+    # If we are processing more than 256 byte, we return an error
+    if serializedHandshakeMessageLen > 256:
+      debug "PayloadV2 malformed: too many public keys contained in the handshake message"
+      raise newException(NoiseMalformedHandshake, "Too many public keys in handshake message")
+
+  # We get the transport message byte length
+  let transportMessageLen = self.transportMessage.len
+
+  # The output payload as in https://rfc.vac.dev/spec/35/. We concatenate all the PayloadV2 as 
+  # payload = ( protocolId || serializedHandshakeMessageLen || serializedHandshakeMessage || transportMessageLen || transportMessage)
+  # We declare it as a byte sequence of length accordingly to the PayloadV2 information read 
+  var payload = newSeqOfCap[byte](1 + # 1 byte for protocol ID             
+                                  1 + # 1 byte for length of serializedHandshakeMessage field
+                                  serializedHandshakeMessageLen + # serializedHandshakeMessageLen bytes for serializedHandshakeMessage
+                                  8 + # 8 bytes for transportMessageLen
+                                  transportMessageLen # transportMessageLen bytes for transportMessage
+                                  )
+  
+  # We concatenate all the data
+  # The protocol ID (1 byte) and handshake message length (1 byte) can be directly casted to byte to allow direct copy to the payload byte sequence
+  payload.add self.protocolId.byte
```

Comment:
This will be done in next PR (https://github.com/status-im/nim-waku/pull/933), when the list of supported protocol IDs and Noise handshakes is introduced.